File: app/src/main/java/com/possible/tourrefactorsample/data/database/BookDataSource.java
package com.possible.tourrefactorsample.data.database;

import android.util.Log;

import com.possible.tourrefactorsample.data.models.Book;
import com.possible.tourrefactorsample.data.models.BookDao;
import com.possible.tourrefactorsample.data.models.DaoSession;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

public class BookDataSource {

    private static final String TAG = BookDataSource.class.getSimpleName();

    private DaoSession daoSession;
    private BookDao bookDao;

    public BookDataSource(DaoSession session) {
        daoSession = session;
        bookDao = daoSession.getBookDao();
    }

    public List<Book> getBooks() {
        List<Book> books = new ArrayList<>();

        books.addAll(bookDao.loadAll());

        return books;
    }

    public void saveBooks(final List<Book> books) {
        try {
            daoSession.callInTx(new Callable<Object>() {
                @Override
                public Object call() throws Exception {
                    removeBooks();
                    insertBooks(books);
                    return null;
                }
            });
        } catch(Exception e) {
            Log.e(TAG, "Unable to update table", e);
        }
    }

    private void removeBooks() {
        daoSession.getBookDao().deleteAll();
    }

    private void insertBooks(List<Book> books) {
        bookDao.insertInTx(books);
    }
}

File: app/src/main/kotlin/com/possible/architecturesample/data/database/BookDataSource.kt
package com.possible.architecturesample.data.database

import android.util.Log
import com.possible.architecturesample.data.models.Book
import com.possible.architecturesample.data.models.BookDao
import com.possible.architecturesample.data.models.DaoSession
import java.util.ArrayList

class BookDataSource(private val daoSession: DaoSession) {
    private val bookDao: BookDao

    init {
        bookDao = daoSession.bookDao
    }

    val books: List<Book>
        get() {
            val books = ArrayList<Book>()

            books.addAll(bookDao.loadAll())

            return books
        }

    fun saveBooks(books: List<Book>) {
        try {
            daoSession.callInTx {
                removeBooks()
                insertBooks(books)
                null
            }
        } catch (e: Exception) {
            Log.e(TAG, "Unable to update table", e)
        }

    }

    private fun removeBooks() {
        daoSession.bookDao.deleteAll()
    }

    private fun insertBooks(books: List<Book>) {
        bookDao.insertInTx(books)
    }

    companion object {
        private val TAG = BookDataSource::class.java.simpleName
    }
}
